src/main/lang/Nix.flex

```diff
@@ -2,53 +2,33 @@ package org.nixos.idea.lang;
 
 import com.intellij.lexer.FlexLexer;
 import com.intellij.psi.tree.IElementType;
-import it.unimi.dsi.fastutil.longs.Long2IntMap;
-import it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap;
-import it.unimi.dsi.fastutil.longs.LongArrayList;
-import it.unimi.dsi.fastutil.longs.LongList;
+import it.unimi.dsi.fastutil.ints.AbstractIntList;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
 
 import static org.nixos.idea.psi.NixTypes.*;
 
 %%
 
 %{
-  private final LongList states = new LongArrayList();
-  private final Long2IntMap stateIndexMap = new Long2IntOpenHashMap();
+  private final AbstractIntList states = new IntArrayList();
 
-  {
-    states.add(YYINITIAL);
-    stateIndexMap.put(YYINITIAL, 0);
+  private void pushState(int newState) {
+      // store current state on the stack to allow restoring it in popState(...)
+      states.push(yystate());
+      yybegin(newState);
   }
 
-  private int currentStateIndex = 0;
-  private int parentStateIndex = 0;
-
-  public int getStateIndex() {
-    return currentStateIndex;
-  }
-
-  public void restoreState(int stateIndex) {
-    long state = states.getLong(stateIndex);
-    currentStateIndex = stateIndex;
-    parentStateIndex = (int) (state >> 32);
-    yybegin((int) state);
-  }
-
-  private void pushState(int yystate) {
-    long state = ((long) currentStateIndex << 32) | ((long) yystate & 0x0FFFFFFFFL);
-    int stateIndex = stateIndexMap.get(state); // Returns 0 if not found
-    if (stateIndex == 0 && state != YYINITIAL) {
-      stateIndex = states.size();
-      states.add(state);
-      stateIndexMap.put(state, stateIndex);
+  private void popState(int expectedState) {
+    // safe-guard, because we always know which state we're currently in in the rules below
+    if (yystate() != expectedState) {
+        throw new IllegalStateException(String.format("Unexpected state. Current: %d, expected: %d", yystate(), expectedState));
     }
-    parentStateIndex = currentStateIndex;
-    currentStateIndex = stateIndex;
-    yybegin(yystate);
+    // start the lexer with the previous state, which was stored by pushState(...)
+    yybegin(states.popInt());
```

Comment:
The stack can be empty if the file is not a valid.

Reply:
fixed by 964aa65afdfed9ed9092ff33cbb0e9c93bdf66eb
Calling `popState` from `YYINITIAL` is a always a bug (I think) and now causes an exception.

Reply:
The method `states.popInt()` would have also caused an exception, but having a separate precondition doesn't hurt.
I would personally have chosen another route, kept a single rule for all three states, and made `popState` a no-op on `YYINITIAL` when the stack is empty. Anyway, your solution of not calling `popState` from `YYINITIAL` (6ac5ab6199f) works as well.